In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import drive
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
train = pd.read_csv('/content/drive/My Drive/recsys-hw2/train.csv')
songs = pd.read_csv('/content/drive/My Drive/recsys-hw2/songs.csv')

In [4]:
train.head(5)

msno  ... target
0  FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  ...      1
1  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...      1
2  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...      1
3  Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  ...      1
4  FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  ...      1

[5 rows x 6 columns]

In [5]:
songs.head(5)

song_id  ...  language
0  CXoTN1eb7AI+DntdU1vbcwGRV4SCIDxZu+YD8JP8r4E=  ...       3.0
1  o0kFgae9QtnYgRkVPqLJwa05zIhRlUjfF7O1tDw0ZDU=  ...      31.0
2  DwVvVurfpuz+XPuFvucclVQEyPqcpUkHR0ne1RQzPs0=  ...      31.0
3  dKMBWoZyScdxSkihKG+Vf47nc18N9q4m58+b4e7dSSE=  ...       3.0
4  W3bqWd3T+VeHFzHAUfARgW9AvVRaF4N5Yzm4Mr6Eo/o=  ...      52.0

[5 rows x 7 columns]

In [6]:
train.drop(['source_system_tab', 'source_screen_name', 'source_type'], axis=1, inplace=True)

train = train[train['target'] == 1]

target = train['target']

train.drop(['target'], axis=1, inplace=True)

In [7]:
temp = train.groupby(['msno'])['song_id'].apply(list)
temp = temp.to_frame().reset_index()

temp.columns = ['msno', 'song_id_history']

histories = list(temp['song_id_history'])

print(len(histories[0]))

293


Для каждого пользователя вытащили историю его лайков, обучим на них skipgram. Взял не очень большие размер эмбеддинга (64) и окно (5), чтобы училось чуть-чуть быстрее.

In [8]:
%%time

from gensim.models import Word2Vec

model = Word2Vec(window=5, sg=1, hs=0, size=64, negative=15, alpha=0.03, min_alpha=0.0007)
model.build_vocab(histories, progress_per=100)

model.train(histories, total_examples=model.corpus_count, epochs=10, report_delay=1)

CPU times: user 15min 39s, sys: 777 ms, total: 15min 39s
Wall time: 5min 25s


In [9]:
print(model[model.wv.vocab].shape)

(55816, 64)


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  """Entry point for launching an IPython kernel.


In [10]:
print(model[model.wv.vocab][0])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  """Entry point for launching an IPython kernel.


[-0.21026215 -0.21907547 -0.24418934 -0.1990588  -0.883448    0.5492157
  0.460374   -0.2067067   0.1746037  -0.11719668  0.97223    -0.02667366
 -0.0433716  -0.08816152 -0.17095849 -0.23399873  0.37236282  0.31661746
  0.0831437   0.10915186  0.14463888  0.44264695 -0.6783294  -0.06961908
  0.5325378  -0.0734586  -0.2005061   0.8451047  -0.40111002 -0.25154147
 -0.18716572  1.1008812   0.03107803 -0.18050763  0.25047576  0.4464444
  0.32369924  0.02690399  0.0949702   0.2885624  -0.43931878  0.3938591
  0.33923978 -0.75368005  0.0539429   0.1678574  -0.98702276 -0.4033384
 -0.04532693 -0.51851124  0.1601011  -0.16902839 -0.28049377 -0.23418777
  0.06987635 -0.02598332  0.28869376 -0.04344971 -0.4473675   0.2624831
 -0.5233734   0.3073005  -0.53402865  0.28218353]


In [11]:
model.init_sims(replace=True)

In [12]:
def get_similars(initial_song_id, k=20):
    similars = model.similar_by_vector(model[initial_song_id], topn=k + 1)
    res = []
    
    for song_id, similarity in similars:
        row = songs[songs['song_id'] == song_id].to_dict('records')[0]
        row['similarity'] = similarity
        res.append(row)
        
    return pd.DataFrame(res)

In [13]:
initial_song_id = list(model.wv.vocab.items())[4000][0]
get_similars(initial_song_id)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: DeprecationWarning: Call to deprecated `similar_by_vector` (Method will be removed in 4.0.0, use self.wv.similar_by_vector() instead).
  
/usr/local/lib/python3.6/dist-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


song_id  ...  similarity
0   7HDcHFAVdw76mL4Z5hbz300+s1oSUstaOfmaLw0+Dfo=  ...    1.000000
1   u9z+kvlYCkrSlDdJv3O/t5h+JpZJrSn2ZBI1GWe+EYQ=  ...    0.860061
2   VFWmrWVR8IbIiP4iwS+RwATCVRqZozRMJ06PqQQeEBI=  ...    0.834578
3   4/f7S7iNOygpoPNxUzMv7jOPLg4HGYhpbEO7i41zUzU=  ...    0.816846
4   J8poi1Mg9pKUq7Kw/w9y7JVmX3AH5OeFh4YJJi3hieY=  ...    0.816453
5   FsI4pS2XGlzWGBGeSonZ8mvxsq3wHjRF19VjE2/elzc=  ...    0.786667
6   uuv/oI/dfOrHOHwDN9co6xyS484UCmel5TPfqOWCiU8=  ...    0.782698
7   2AEXIjiK5meckRxl3k3zv5FnpimIvBrtgsI6FyFOlFI=  ...    0.779018
8   ES86JRfxcI9nxEfNv70Y772KQc9gxpykvAD7XRTsBgM=  ...    0.778792
9   lhrN5Q069mAF35yPS+P4Rk1c1L4jhrAt3T2UjvT8yXc=  ...    0.773167
10  CBFwS48p3EnalLJxLaC5chzCpEkUIyNvyXKL0OvD1X4=  ...    0.770224
11  doA/FLODmvznVgfC16Apl98I3asYPyGJroyCuK7O0eg=  ...    0.768307
12  e8w6DjV+e6p3LigDIYIMwQoldbpUSTk06wHSZnbW+8o=  ...    0.767259
13  ZCNuPSqqMMOybxfQUNNPz5p9eNeHs+BIsfheAYw1rrA=  ...    0.766770
14  oqWJlYOCx9z2e7eTgKfq2VDAMaFDJYeHdaM4HsgWBE8=  ...    0.764637
15  qm1G/0r9ZHsHskGuhIBOqKl1emPSQMAiVARG49LfMLI=  ...    0.764091
16  WiBDGcHgmrkJ1PiTlYqBFlVpDgaoEJtn8PgCJnlaWaY=  ...    0.762592
17  6HYyGPK6J1jG3CJTw4tjBL0gTAZ/XxCt757pwy1+tcw=  ...    0.762017
18  di8YABFd7GMSOEM4NGFkADekD7AY8eLN1qYTAYSkd/s=  ...    0.760359
19  udsP+7qJrwSlKCcEgeEv3ol5yMmi/Xaa1BYVZfG6z0o=  ...    0.759849
20  eyQjj8u75mf0vmmJWicH065PbT+152++Auy3Lh/V6g8=  ...    0.759272

[21 rows x 8 columns]

Посмотрим на симилары. К Рианне советуется разнообразная попса (включая саму Рианну). Уже неплохо.

In [14]:
song_embeddings = pd.DataFrame(model[model.wv.vocab])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  """Entry point for launching an IPython kernel.


In [15]:
song_embeddings['song_id'] = np.array(list(model.wv.vocab.keys()))

In [ ]:
# song_embeddings.to_csv('/content/drive/My Drive/recsys-hw2/song_embeddings.csv')

In [16]:
song_embeddings.set_index('song_id', inplace=True)

song_embeddings.head(5)

0   ...        63
song_id                                                 ...          
aRNy4rKgWyDwTjsMIcAbtpGvddDZ7L7OGUeJ09LAM2w= -0.065482  ...  0.087881
J4qKkLIoW7aYACuTupHLAPZYmRp08en1AEux+GSUzdw=  0.123072  ...  0.047978
zHqZ07gn+YvF36FWzv9+y8KiCMhYhdAUS+vSIKY3UZY=  0.137898  ...  0.038334
v/3onppBGoSpGsWb8iaCIO8eX5+iacbH5a4ZUhT7N54=  0.108986  ... -0.083026
upH6pOAUd+iV/MkpzeELvqEFoTEIVhsV9eML8N7/gUM= -0.006570  ...  0.054631

[5 rows x 64 columns]

Посчитаем эмбеддинг юзера как среднее арифметическое эмбеддингов послушанных им песен.

In [17]:
%%time

from collections import defaultdict

def get_user_embeddings(df):
    user_embeddings = defaultdict(lambda: np.zeros(64))
    user_counts = defaultdict(int)

    for (_, row) in df.iterrows():
        user_id = row['msno']
        song_id = row['song_id']

        if song_id in model.wv:
            user_embeddings[user_id] += model.wv[song_id]
            user_counts[user_id] += 1

    for user_id in user_embeddings.keys():
        user_embeddings[user_id] /= user_counts[user_id]

    return user_embeddings

user_embeddings = get_user_embeddings(train)

CPU times: user 5min 56s, sys: 4.07 s, total: 6min
Wall time: 5min 54s


In [18]:
user_embeddings = pd.DataFrame(user_embeddings)
user_embeddings = user_embeddings.T

user_embeddings.head(5)

0   ...        63
FGtllVqz18RPiwJj/edr2gV78zirAiY/9SmYvia+kCg=  0.069901  ...  0.065408
Xumu+NIjS6QYVxDS4/t3SawvJ7viT9hPKXmf0RtLNx8=  0.070406  ...  0.040430
uHqAtShXTRXju5GE8ri3ITsVFepPf8jUoCF7ffNOuqE=  0.084202  ...  0.012250
3g0bC24RD7QUeALY1IF/e8L6X3fHG298ZFIsCEG9T9Y=  0.136235  ...  0.029513
TJU0Gfvy7FB+r89bWovPKXTjuApTCiv3xg/tt5shR78=  0.099432  ...  0.021507

[5 rows x 64 columns]

In [ ]:
# res.to_csv('/content/drive/My Drive/recsys-hw2/user_embeddings.csv')

Сохранил эмбеддинги в .csv, чтобы использовать их в ноутбуке с градиентным бустингом.

## Посчитаем AUC

In [20]:
train = pd.read_csv('/content/drive/My Drive/recsys-hw2/train.csv')

train.drop(['source_system_tab', 'source_screen_name', 'source_type'], axis=1, inplace=True)

target = train['target']

train.drop(['target'], axis=1, inplace=True)

In [21]:
MAX_SIZE = 1000000

train = train[:MAX_SIZE]
target = target[:MAX_SIZE]

In [22]:
def get_dot_product(song_id, user_id):
  if song_id not in song_embeddings.index:
    return 0
  if user_id not in user_embeddings.index:
    return 0
  
  return (np.sum(user_embeddings.loc[user_id] * song_embeddings.loc[song_id]))

In [24]:
%%time

train['dot_product'] = train.apply(lambda row: get_dot_product(row['song_id'], row['msno']), axis = 1)

CPU times: user 8min 56s, sys: 10.3 s, total: 9min 6s
Wall time: 8min 50s


In [27]:
from sklearn.metrics import roc_auc_score

score = roc_auc_score(target, train['dot_product'])

print(score)

0.6686055479653281


Ожидаемо, roc_auc = 0.66 оказался слабее чем у простого катбуста (0.75) и катбуста с эмбеддингами (0.86)